Explore a small gene expression dataset
========================================================

```{r include = FALSE}
## I format my code intentionally!
## do not re-format it for me!
opts_chunk$set(tidy = FALSE)

## toggle to turn solution chunks on/off
##opts_chunk$set(eval = FALSE, echo = FALSE)
```

### Optional getting started advice

*Ignore if you don't need this bit of support.*

This is one in a series of tutorials in which we work with the `photoRec` dataset. Now is the time to make sure you are working in an appropriate directory on your computer, probably through the use of an [RStudio project](http://www.stat.ubc.ca/~jenny/STAT545A/block01_basicsWorkspaceWorkingDirProject.html). To ensure a clean slate, you may wish to clean out your workspace and restart R (both available from the RStudio Session menu, among other methods). Confirm that the new R process has the desired working directory, for example, with the `getwd()` command or by glancing at the top of RStudio's Console pane.

Open a new R script (in RStudio, File > New > R Script). Develop and run your code from there (recommended) or periodicially copy "good" commands from the history. In due course, save this script with a name ending in `.r` or `.R`, __containing no spaces or other funny stuff__, and evoking "seminar01" and "data exploration".

See [R basics, workspace and working directory, RStudio projects](http://www.stat.ubc.ca/~jenny/STAT545A/block01_basicsWorkspaceWorkingDirProject.html) if you need some help developing your basic R mojo.

### Obtain the data file `GSE4051_MINI.txt`

We will work with a small data excerpt. Please save this file locally, for example, in the directory associated with your RStudio Project:

  * <http://www.ugrad.stat.ubc.ca/~stat540/examples/photoRec/data/GSE4051_MINI.txt>
  
### The `photoRec` data

*to be polished*

See the [`photoRec` README](https://github.com/jennybc/stat540_2014/tree/master/examples/photoRec).

Gene expression data from photo receptor cells in mice. Various developmental stages, two genotypes.

The excerpt we work with here contains gene expression data for 3 genes. Their arcane names have been replaced with 3 randomly selected Pokemon attacks: crabHammer, eggBomb, and poisonFang.

### Load `GSE4051_MINI.txt`

```{r include = FALSE}
## here is how we actually load the data
## reflects directory structure in 2014_STAT540-devel and 2014_STAT540
## handles working directory = that of file or at top-level of repo
pwd <- getwd()
if(basename(pwd) %in% c("seminars_2014", "seminars")) {
  prefix <- ".."
  } else {
    prefix <- ""
    }
prDat <- read.table(file.path(prefix, "examples/photoRec/data/GSE4051_MINI.txt"),
                              header = TRUE, row.names = 1)
```

We load the mini gene expression dataset into a `data.frame`, which is the preferred way to handle tabular data in R.

If you have not already done so, now is a good time to work through [Basic care and feeding of data in R](http://www.stat.ubc.ca/~jenny/STAT545A/block02_careFeedingData.html). The exercises below will test whether your basic understanding of `data.frames` is developed enough to explore the mini gene expression data we work with here.

If you are new to R, you'll want to work through [R objects (beyond data.frames) and indexing](http://www.stat.ubc.ca/~jenny/STAT545A/block03_basicObjects.html) at some in the near future as well.

Read the data from a local file in the current working directory and use `str()` to get a basic overview of the resulting `prDat` object:

```{r eval = FALSE}
prDat <- read.table("GSE4051_MINI.txt", header = TRUE, row.names = 1)
```

Does the `read.table()` command fail for you? If so, have you taken control of your working directory? Have you placed `GSE4051_MINI.txt` in that directory or edited the `read.table()` command to reflect its location? Mastering the working directory and filepaths is a surprisingly important part of the learning curve. Submit commands like these to do some troubleshooting:

```{r eval = FALSE}
getwd()      # is working directory what you think it should be?
list.files() # do see GSE4051_MINI.txt sitting there?!?
setwd()      # use with moderation
```

Upon successful import, `str()` the object as a sanity check:
```{r}
str(prDat)
```

Exercise: Why do we use the `header =` and `row.names =` arguments above upon import? Submit the command *without* these arguments and note any difference in the result. Form the habit of reading error messages *carefully* and working the problem. Mastering the arguments of `read.table()` and friends is time well spent.


```{r eval = FALSE, include = FALSE}
## currently deactivated but shows reading straight from github
library(RCurl)
x <- getURL("https://raw.github.com/jennybc/stat540_2014/master/examples/photoRec/data/GSE4051_MINI.txt")
prDat <- read.table(text = x, header = TRUE, row.names = 1)
```

```{r eval = FALSE, include = FALSE}
## currently deactivated but shows reading straight from course website
prURL <- "http://www.ugrad.stat.ubc.ca/~stat540/examples/photoRec/data/GSE4051_MINI.txt"
prDat <- read.table(file = prURL, header = TRUE, row.names = 1)
```

### Basic exploration of `prDat`

Use R to answer these questions. You could create a script with these questions as comments, then insert actual code to get the answers. You could use RStudio's "Compile Notebook" command to create a nice little report!

How many rows are there? Hint: `nrow()`, `dim()`.

How many columns or variables are there? Hint: `ncol()`, `length()`, `dim()`.

Inspect the first few observations or the last few or a random sample. Hint: `head()`, `tail()`, `x[i, j]` combined with `sample()`.

What does row correspond to -- different genes or different mice?

What are the variable names? Hint: `names()`, `dimnames()`.

What "flavor" is each variable, i.e. numeric, character, factor? Hint: `str()`.

For `sample`, do a sanity check that each integer between 1 and the number of rows in the dataset occurs exactly once. Hint: `a:b`,  `seq()`, `seq_len()`, `sort()`, `table()`, `==`, `all()`, `all.equal()`, `identical()`.

For each factor variable, what are the levels? Hint: `levels()`, `str()`.

How many observations do we have for each level of `devStage`? For `gType`? Hint: `summary()`, `table()`.

Perform a cross-tabulation of `devStage` and `gType`. Hint: `table()`.

If you had to take a wild guess, what do you think the *intended* experimental design was? What actually happened in real life?

For each quantitative variable, what are the extremes? How about average or median? Hint: `min()`, `max()`, `range()`, `summary()`, `fivenum()`, `mean()`, `median()`, `quantile()`.

```{r}
str(prDat)
dim(prDat)
nrow(prDat)
ncol(prDat)
length(prDat)
names(prDat)
head(prDat)
tail(prDat)
prDat[sample(nrow(prDat), size = 6), ]
table(table(prDat$sample))
all(sort(prDat$sample) == seq_len(nrow(prDat)))
identical(sort(prDat$sample), seq_len(nrow(prDat)))
str(prDat$devStage)
summary(prDat$devStage)
table(prDat$devStage)
levels(prDat$devStage)
summary(prDat$gType)
table(prDat$gType)
levels(prDat$gType)
table(prDat$devStage, prDat$gType)
with(prDat, table(devStage, gType))
min(prDat$crabHammer)
max(prDat$eggBomb)
range(prDat$poisonFang)
summary(prDat$crabHammer)
fivenum(prDat$eggBomb)
mean(prDat$poisonFang)
median(prDat$crabHammer)
quantile(prDat$crabHammer, probs = 0.5)
quantile(prDat$crabHammer, probs = c(0.05, 0.10, 0.9, 0.95))
```

### Indexing and subsetting `prDat`

You'll definitely need the knowledge in [R objects (beyond data.frames) and indexing](http://www.stat.ubc.ca/~jenny/STAT545A/block03_basicObjects.html) now, so go work through it if you need to.

Create a new data.frame called `weeDat` only containing observations for which expression of poisonFang is above 7.5.

For how many observations poisonFang > 7.5? How do they break down by genotype and developmental stage? 

```{r}
weeDat <- subset(prDat, poisonFang > 7.5)
nrow(weeDat)
table(weeDat$gType)
table(weeDat$devStage)
addmargins(with(weeDat, table(devStage, gType)))
```

Print the observations with row names "Sample_16" and "Sample_38" to screen, showing only the 3 gene expression variables.

Which samples have expression of eggBomb less than the 0.10 quantile?

```{r}
prDat[c("Sample_16", "Sample_38"), c("crabHammer", "eggBomb", "poisonFang")]
rownames(prDat[prDat$eggBomb < quantile(prDat$eggBomb, 0.1), ])
```

### Answers!

Code for the exercises is shown in the source document that makes this page, ~~but is hidden here~~. Go read [the R Markdown source](https://github.com/jennybc/stat540_2014/blob/master/seminars/seminar01_basic-data-exploration.rmd) if you want to see it.